src/Algorithms/DirectSearch/CMA.cpp

```diff
@@ -312,7 +338,89 @@ void CMA::updatePopulation( std::vector<IndividualType> const& offspring ) {
 void CMA::step(ObjectiveFunctionType const& function){
 	std::vector<IndividualType> offspring = generateOffspring();
 	PenalizingEvaluator penalizingEvaluator;
+	penalizingEvaluator.m_numEvaluations = m_numEvaluations;
 	penalizingEvaluator( function, offspring.begin(), offspring.end() );
+	//check if the number of Evaluations must be increased on a noisy function
+	if(function.isNoisy()){
+		//compute number of points to reevaluate
+		double reevalFraction = m_rLambda * m_lambda;
+		std::size_t lambdaReeval = m_rLambda * m_lambda;
+		double rest = reevalFraction - lambdaReeval;
+		if(rest > 0){
+			lambdaReeval += coinToss(*mpe_rng,rest);
+		}
+		//only continue if we have at elast one point
+		if(lambdaReeval > 0){
+			//save old function values of the population
+			std::vector<shark::KeyValuePair<double,std::size_t> > ranksOld(m_lambda);
+			for(std::size_t i = 0; i != m_lambda; ++i){
+				ranksOld[i].key = offspring[i].penalizedFitness();
+				ranksOld[i].value = i;
+			}
+			
+			//compute and save new function values of the population
+			penalizingEvaluator(function,offspring.begin(),offspring.begin() + lambdaReeval);
+			std::vector<shark::KeyValuePair<double,std::size_t> > ranksNew(m_lambda);
+			for(std::size_t i = 0; i != m_lambda; ++i){
+				ranksNew[i].key = offspring[i].penalizedFitness();
+				ranksNew[i].value = i;
+			}
+			
+			// update function values of the population to be the mean of old and new values,
```

Comment:
This comment should probably contain a reference to the paper. The reference could also go into the class doxygen header.